vmips-mips-emulator/src/doc/vmipstool.1
.\" Automatically generated by Pod::Man 2.27 (Pod::Simple 3.28)
.\"
.\" Standard preamble:
.\" ========================================================================
.de Sp \" Vertical space (when we can't use .PP)
.if t .sp .5v
.if n .sp
..
.de Vb \" Begin verbatim text
.ft CW
.nf
.ne \\$1
..
.de Ve \" End verbatim text
.ft R
.fi
..
.\" Set up some character translations and predefined strings.  \*(-- will
.\" give an unbreakable dash, \*(PI will give pi, \*(L" will give a left
.\" double quote, and \*(R" will give a right double quote.  \*(C+ will
.\" give a nicer C++.  Capital omega is used to do unbreakable dashes and
.\" therefore won't be available.  \*(C` and \*(C' expand to `' in nroff,
.\" nothing in troff, for use with C<>.
.tr \(*W-
.ds C+ C\v'-.1v'\h'-1p'\s-2+\h'-1p'+\s0\v'.1v'\h'-1p'
.ie n \{\
.    ds -- \(*W-
.    ds PI pi
.    if (\n(.H=4u)&(1m=24u) .ds -- \(*W\h'-12u'\(*W\h'-12u'-\" diablo 10 pitch
.    if (\n(.H=4u)&(1m=20u) .ds -- \(*W\h'-12u'\(*W\h'-8u'-\"  diablo 12 pitch
.    ds L" ""
.    ds R" ""
.    ds C` ""
.    ds C' ""
'br\}
.el\{\
.    ds -- \|\(em\|
.    ds PI \(*p
.    ds L" ``
.    ds R" ''
.    ds C`
.    ds C'
'br\}
.\"
.\" Escape single quotes in literal strings from groff's Unicode transform.
.ie \n(.g .ds Aq \(aq
.el       .ds Aq '
.\"
.\" If the F register is turned on, we'll generate index entries on stderr for
.\" titles (.TH), headers (.SH), subsections (.SS), items (.Ip), and index
.\" entries marked with X<> in POD.  Of course, you'll have to process the
.\" output yourself in some meaningful fashion.
.\"
.\" Avoid warning from groff about undefined register 'F'.
.de IX
..
.nr rF 0
.if \n(.g .if rF .nr rF 1
.if (\n(rF:(\n(.g==0)) \{
.    if \nF \{
.        de IX
.        tm Index:\\$1\t\\n%\t"\\$2"
..
.        if !\nF==2 \{
.            nr % 0
.            nr F 2
.        \}
.    \}
.\}
.rr rF
.\"
.\" Accent mark definitions (@(#)ms.acc 1.5 88/02/08 SMI; from UCB 4.2).
.\" Fear.  Run.  Save yourself.  No user-serviceable parts.
.    \" fudge factors for nroff and troff
.if n \{\
.    ds #H 0
.    ds #V .8m
.    ds #F .3m
.    ds #[ \f1
.    ds #] \fP
.\}
.if t \{\
.    ds #H ((1u-(\\\\n(.fu%2u))*.13m)
.    ds #V .6m
.    ds #F 0
.    ds #[ \&
.    ds #] \&
.\}
.    \" simple accents for nroff and troff
.if n \{\
.    ds ' \&
.    ds ` \&
.    ds ^ \&
.    ds , \&
.    ds ~ ~
.    ds /
.\}
.if t \{\
.    ds ' \\k:\h'-(\\n(.wu*8/10-\*(#H)'\'\h"|\\n:u"
.    ds ` \\k:\h'-(\\n(.wu*8/10-\*(#H)'\`\h'|\\n:u'
.    ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'^\h'|\\n:u'
.    ds , \\k:\h'-(\\n(.wu*8/10)',\h'|\\n:u'
.    ds ~ \\k:\h'-(\\n(.wu-\*(#H-.1m)'~\h'|\\n:u'
.    ds / \\k:\h'-(\\n(.wu*8/10-\*(#H)'\z\(sl\h'|\\n:u'
.\}
.    \" troff and (daisy-wheel) nroff accents
.ds : \\k:\h'-(\\n(.wu*8/10-\*(#H+.1m+\*(#F)'\v'-\*(#V'\z.\h'.2m+\*(#F'.\h'|\\n:u'\v'\*(#V'
.ds 8 \h'\*(#H'\(*b\h'-\*(#H'
.ds o \\k:\h'-(\\n(.wu+\w'\(de'u-\*(#H)/2u'\v'-.3n'\*(#[\z\(de\v'.3n'\h'|\\n:u'\*(#]
.ds d- \h'\*(#H'\(pd\h'-\w'~'u'\v'-.25m'\f2\(hy\fP\v'.25m'\h'-\*(#H'
.ds D- D\\k:\h'-\w'D'u'\v'-.11m'\z\(hy\v'.11m'\h'|\\n:u'
.ds th \*(#[\v'.3m'\s+1I\s-1\v'-.3m'\h'-(\w'I'u*2/3)'\s-1o\s+1\*(#]
.ds Th \*(#[\s+2I\s-2\h'-\w'I'u*3/5'\v'-.3m'o\v'.3m'\*(#]
.ds ae a\h'-(\w'a'u*4/10)'e
.ds Ae A\h'-(\w'A'u*4/10)'E
.    \" corrections for vroff
.if v .ds ~ \\k:\h'-(\\n(.wu*9/10-\*(#H)'\s-2\u~\d\s+2\h'|\\n:u'
.if v .ds ^ \\k:\h'-(\\n(.wu*10/11-\*(#H)'\v'-.4m'^\v'.4m'\h'|\\n:u'
.    \" for low resolution devices (crt and lpr)
.if \n(.H>23 .if \n(.V>19 \
\{\
.    ds : e
.    ds 8 ss
.    ds o a
.    ds d- d\h'-1'\(ga
.    ds D- D\h'-1'\(hy
.    ds th \o'bp'
.    ds Th \o'LP'
.    ds ae ae
.    ds Ae AE
.\}
.rm #[ #] #H #V #F C
.\" ========================================================================
.\"
.IX Title "VMIPSTOOL 1"
.TH VMIPSTOOL 1 "2018-08-20" "vmips 1.5.1" "VMIPS Programmer's Manual"
.\" For nroff, turn off justification.  Always turn off hyphenation; it makes
.\" way too many mistakes in technical documents.
.if n .ad l
.nh
.SH "NAME"
vmipstool \- VMIPS front end for common programming tasks
.SH "SYNOPSIS"
.IX Header "SYNOPSIS"
.Vb 10
\&        vmipstool [ \-\-verbose ] [ \-\-dry\-run ] \-\-compile [ FLAGS ]
\&            FILE.c \-o FILE.o
\&        vmipstool [ \-\-verbose ] [ \-\-dry\-run ] \-\-preprocess [ FLAGS ] FILE
\&        vmipstool [ \-\-verbose ] [ \-\-dry\-run ] \-\-assemble [ FLAGS ]
\&            FILE.s \-o FILE.o
\&        vmipstool [ \-\-verbose ] [ \-\-dry\-run ] [ \-\-ld\-script=T ] \-\-link
\&            [ FLAGS ] FILE1.o ... FILEn.o \-o PROG
\&        vmipstool [ \-\-verbose ] [ \-\-dry\-run ] \-\-make\-rom PROG PROG.rom
\&        vmipstool [ \-\-verbose ] [ \-\-dry\-run ] \-\-disassemble\-rom PROG.rom
\&        vmipstool [ \-\-verbose ] [ \-\-dry\-run ] \-\-disassemble\-word PC INSTR
\&        vmipstool [ \-\-verbose ] [ \-\-dry\-run ] \-\-disassemble PROG (or FILE.o)
\&        vmipstool [ \-\-verbose ] [ \-\-dry\-run ] \-\-swap\-words INPUT OUTPUT
\&        vmipstool \-\-help
\&        vmipstool \-\-version
.Ve
.SH "DESCRIPTION"
.IX Header "DESCRIPTION"
\&\fBvmipstool\fR is intended to be a friendly front-end to the process
of compiling, linking, and assembling code for \s-1VMIPS\s0 using the \s-1GNU\s0 Compiler
Collection (\s-1GCC\s0) and \s-1GNU\s0 Binutils.
.PP
Note that you do not need to use \fBvmipstool\fR, or even \s-1GCC,\s0 to compile
programs for \s-1VMIPS\s0; you can use any \s-1MIPS\s0 compiler and assembler you have handy.
.SH "OPTIONS"
.IX Header "OPTIONS"
This is what the different command line options mean:
.IP "\fB\-\-help\fR" 4
.IX Item "--help"
Display this help message and exit.
.IP "\fB\-\-version\fR" 4
.IX Item "--version"
Display the version of vmipstool and exit.
.IP "\fB\-\-verbose\fR" 4
.IX Item "--verbose"
Echo commands as they are run.
.IP "\fB\-\-dry\-run\fR" 4
.IX Item "--dry-run"
Don't actually run anything; use with \-\-verbose.
.IP "\fB\-\-ld\-script=T\fR" 4
.IX Item "--ld-script=T"
Use T as the linker script (instead of default script); use with \-\-link.
.IP "\fB\-\-compile\fR" 4
.IX Item "--compile"
Compile C code.  The remainder of the command   
line must consist of arguments to the \s-1GNU C\s0 compiler.
.IP "\fB\-\-preprocess\fR" 4
.IX Item "--preprocess"
Preprocess C source code or assembly code. The remainder of the command   
line must consist of arguments to the \s-1GNU C\s0 preprocessor.
.IP "\fB\-\-assemble\fR" 4
.IX Item "--assemble"
Translate assembly code to object files. The remainder of the command   
line must consist of arguments to the \s-1GNU\s0 assembler.
.IP "\fB\-\-link\fR" 4
.IX Item "--link"
Link objects together to create an executable. The remainder of the command
line must consist of arguments to the \s-1GNU\s0 linker.
.IP "\fB\-\-make\-rom\fR" 4
.IX Item "--make-rom"
Write a program into a \s-1ROM\s0 file. The next 2 arguments are the executable and
the \s-1ROM\s0 file, respectively.
.IP "\fB\-\-disassemble\fR" 4
.IX Item "--disassemble"
Disassemble a relocatable object file (*.o file) or an executable.
.IP "\fB\-\-disassemble\-rom\fR" 4
.IX Item "--disassemble-rom"
Disassemble arbitrary data, possibly including \s-1ROM\s0 files. (More
information is available with \-\-disassemble, but it only works on programs
which have not been written into ROMs.)
.IP "\fB\-\-disassemble\-word\fR" 4
.IX Item "--disassemble-word"
Disassemble an instruction whose binary encoding is passed as a command-line
argument. Because instructions can have PC-relative immediate arguments, the
\&\s-1PC\s0 must also be specified.
.IP "\fB\-\-swap\-words\fR" 4
.IX Item "--swap-words"
Copy the input file to the output file, reversing the byte-order of each
32\-bit word in the process. This can be used to translate chunks of data
from big-endian to little-endian, or vice-versa.
.SH "FILES"
.IX Header "FILES"
Note that \fBvmipstool\fR consults your \fI/etc/vmipsrc\fR and
\&\fI~/.vmipsrc\fR to determine where the \s-1MIPS\s0 cross compiler, assembler,
and \fBobjdump\fR and \fBobjcopy\fR tools are.  If you install new
cross-tools, you should edit these configuration files to reflect the new
location of the \s-1MIPS\s0 tools, and to reflect their default endianness, by
changing the settings of \fBmipstoolprefix\fR and \fBbigendian\fR.
.SH "BUGS"
.IX Header "BUGS"
For instructions on reporting bugs, see
the \*(L"Reporting Bugs\*(R" appendix of the Info manual.
.SH "SEE ALSO"
.IX Header "SEE ALSO"
\&\fIvmips\fR\|(1), \fIas\fR\|(1), \fIld\fR\|(1), \fIgdb\fR\|(1), 
and the Info entries for \fIgcc\fR, \fIas\fR,
\&\fIld\fR, \fIbinutils\fR and \fIgdb\fR.
.SH "AUTHOR"
.IX Header "AUTHOR"
vmipstool was written by Brian Gaeke.
.SH "COPYRIGHT"
.IX Header "COPYRIGHT"
Copyright (c) 2001, 2002, 2004, 2009, 2014 Brian R. Gaeke.
.PP
Permission is hereby granted, free of charge, to any person obtaining a
copy of this document (the \*(L"Document\*(R"), to deal in the Document without
restriction, including without limitation the rights to use, copy,
modify, merge, publish, distribute, sublicense, and/or sell copies of
the Document, and to permit persons to whom the Document is furnished
to do so, subject to the following conditions:
.PP
The above copyright notice and this permission notice shall be included
in all copies or substantial portions of the Document.
.PP
\&\s-1THE DOCUMENT IS PROVIDED \*(L"AS IS\*(R", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE DOCUMENT OR THE USE OR OTHER DEALINGS IN THE
DOCUMENT.\s0


vmips-mips-emulator/src/vmipstool.cc
/* A sensible interface to MIPS cross compilation tools.
   Copyright 2001, 2003, 2004 Brian R. Gaeke.

This file is part of VMIPS.

VMIPS is free software; you can redistribute it and/or modify it
under the terms of the GNU General Public License as published by the
Free Software Foundation; either version 2 of the License, or (at your
option) any later version.

VMIPS is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of MERCHANTABILITY
or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with VMIPS; if not, write to the Free Software Foundation, Inc.,
51 Franklin Street, Fifth Floor, Boston, MA 02110-1301, USA.  */

/* For later: ways to get page size:
long sysconf(_SC_PAGE_SIZE);
or 
long sysconf(_SC_PAGESIZE);
or
int getpagesize(void);
*/

#include "endiantest.h"
#include "fileutils.h"
#include "stub-dis.h"
#include "error.h"
#include "options.h"
#include <cerrno>
#include <cstdarg>
#include <climits>
#include <cstdlib>
#include <cstring>
#include <cstdio>
#include <iostream>
#include <sys/stat.h>
#include <sys/wait.h>
#include <unistd.h>
#include <vector>

const unsigned int MAXARGV = 500;
const unsigned int TMPFNAMESIZE = 26;

int verbose = 0;
int dryrun = 0;

char pkgdatadir[PATH_MAX];
char ccname[PATH_MAX];
char ldname[PATH_MAX];
char objcopyname[PATH_MAX];
char objdumpname[PATH_MAX];
char endianflag[4];

bool program_available (const char *execname) {
  struct stat st;
  if (stat (execname, &st) < 0)
    return false;
  if (!(st.st_mode & S_IXUSR))
    return false;
  return true;
}

int config_error (const char *programname, const char *execname) {
	fprintf (stderr, "vmipstool: Couldn't find MIPS tool '%s' in '%s' "
		"(installation error?)\n", programname, execname);
	return 1;
}

void setup_paths (bool bigendian, const char *mipstoolprefix) {
	strcpy (pkgdatadir, PKGDATADIR);
    // Allow user (or a shell script) to override the path from which the
    // linker script is read. This allows vmipstool to be executed from the
    // source tree without vmips being installed.
    char *user_supplied_pkgdatadir;
    user_supplied_pkgdatadir = getenv ("VMIPS_PKGDATADIR");
    if (user_supplied_pkgdatadir) {
      strncpy (pkgdatadir, user_supplied_pkgdatadir, PATH_MAX);
      if (strlen (user_supplied_pkgdatadir) >= PATH_MAX)
        pkgdatadir[PATH_MAX - 1] = '\0';
    }

	strcpy (ccname, mipstoolprefix);
	strcat (ccname, "gcc");
	strcpy (ldname, mipstoolprefix);
	strcat (ldname, "ld");
	strcpy (objcopyname, mipstoolprefix);
	strcat (objcopyname, "objcopy");
	strcpy (objdumpname, mipstoolprefix);
	strcat (objdumpname, "objdump");

    // Decide which flag should be used to force the endianness of an
    // object file.
    if (bigendian) {
		strcpy (endianflag, "-EB");
    } else {
		strcpy (endianflag, "-EL");
    }
}

/* ROMs are padded to a multiple of this many bytes. I don't think this
   is strictly necessary and will probably go away eventually. */
long pagesz = 4096;

/********************************* Various utils ******************************/

void usage(void) {
	puts("usage:");
	puts(" vmipstool [ VT-FLAGS ] --compile [ FLAGS ] FILE.c -o FILE.o");
	puts(" vmipstool [ VT-FLAGS ] --preprocess [ FLAGS ] FILE");
	puts(" vmipstool [ VT-FLAGS ] --assemble [ FLAGS ] FILE.s -o FILE.o");
	puts
		(" vmipstool [ VT-FLAGS ] --link [ FLAGS ] FILE1.o ... FILEn.o -o PROG");
	puts(" vmipstool [ VT-FLAGS ] --make-rom PROG PROG.rom");
	puts(" vmipstool [ VT-FLAGS ] --disassemble-rom PROG.rom");
	puts(" vmipstool [ VT-FLAGS ] --disassemble PROG (or FILE.o)");
	puts(" vmipstool [ VT-FLAGS ] --swap-words INPUT OUTPUT");
	puts("");
	puts("VT-FLAGS may include:");
	puts(" --help         Display this help message and exit.");
	puts(" --version      Display the version of vmipstool and exit.");
	puts(" --verbose      Echo commands as they are run.");
	puts
		(" --dry-run      Don't actually run anything; use with --verbose.");
	puts(" --ld-script=T  Use T as ld script (instead of default script);");
	puts("                use with --link.");
	puts("");
	puts("Report bugs to <vmips@dgate.org>.");
}

void maybe_echo(char **newargv) {
	int i;
	if (verbose) {
		for (i = 0; newargv[i] != NULL; i++)
			fprintf(stderr, "%s ", newargv[i]);
		fprintf(stderr, "\n");
	}
}

int maybe_run(char **newargv) {
	pid_t pid;
	int status, error = 0;

	if (!dryrun) {
		if ((pid = fork()) == 0) {
			execv(newargv[0], newargv);
			perror (newargv[0]);
			exit (255);
		} else {
			waitpid(pid, &status, 0);
			if (WIFEXITED(status))
				error = (WEXITSTATUS(status) != 0);
			else
				error = 1;
		}
	}
	return error;
}

int echo_and_run_l(char *c, ...) {
	va_list ap;
	char *newargv[MAXARGV];
	int newargc = 0;

	va_start(ap, c);
	newargv[newargc++] = c;
	while (c != NULL) {
		c = va_arg(ap, char *);
		newargv[newargc++] = c;
	}
	va_end(ap);

	maybe_echo(newargv);
	return maybe_run(newargv);
}

int echo_and_run_lv(char *c, ...) {
	va_list ap;
	char *newargv[MAXARGV];
	char **originalargv;
	int newargc = 0;
	int i;

	va_start(ap, c);
	newargv[newargc++] = c;
	while (c != NULL) {
		c = va_arg(ap, char *);
		newargv[newargc++] = c;
	}
	newargc--;
	originalargv = va_arg(ap, char **);
	for (i = 0; originalargv[i] != NULL; i++)
		newargv[newargc++] = originalargv[i];
	newargv[newargc++] = NULL;
	va_end(ap);

	maybe_echo(newargv);
	return maybe_run(newargv);
}

/* Search colon-separated PATH for file named TARGET. Copy found file's
   full pathname into RESULT and return true if found. Returns false and
   leaves result untouched if not found. */
bool search_path(char *result, char *path, const char *target) {
	const char *delims = ":";
	char *dir = strtok (path, delims);
	while (dir != NULL) {
		char trypath[PATH_MAX];
		sprintf (trypath, "%s/%s", dir, target);
		if (can_read_file (trypath)) {
			strcpy (result, trypath);
			return true;
		} else {
			dir = strtok (NULL, delims);
		}
	}
	return false;
}

/********************************* Linking ******************************/

char ldscript_full_path[PATH_MAX];
bool ldscript_error = false;

int can_read_default_ldscript () {
	/* Build the search path */
	char *ldscript_search_path;

	/* so that it can work without being installed */
	char rest_of_path[] =
		":.:..:./sample_code:../sample_code:../../sample_code";
	const char ldscript_name[] = "ld.script";

	ldscript_search_path =
		new char[strlen(rest_of_path) + strlen(pkgdatadir) + 2];
	strcpy(ldscript_search_path, pkgdatadir);
	strcat(ldscript_search_path, rest_of_path);

	/* Look for the ld.script */
	return search_path(ldscript_full_path, ldscript_search_path,
					   ldscript_name);
}

int do_link (int argc, char **argv) {
	if (ldscript_error) {
		fprintf(stderr, "vmipstool: can't access linker script\n");
		return 1;
	} 
	if (!program_available (ldname))
		return config_error ("ld", ldname);
	return echo_and_run_lv (ldname, endianflag, "-T", ldscript_full_path, NULL,
                            argv, NULL);
}

/********************************* ROM-making ******************************/

int copy_with_padded_blocks (char *in, char *out, long size) {
	char *buff = new char[size];
	FILE *f = fopen(in, "rb");
	if (!f) {
		perror(in);
		return -1;
	}
	FILE *g = fopen(out, "wb");
	if (!g) {
		perror(out);
		return -1;
	}
	errno = 0;
	int readcount;
	while ((readcount = fread(buff, 1, size, f)) != 0) {
		if (readcount < size) {
			if (errno) {
				perror(in);
				fclose(f);
				fclose(g);
				return -1;
			} else {
				for (int i = readcount; i < size; i++)
					buff[i] = '\0';
			}
		}
		int writecount = fwrite(buff, 1, size, g);
		if (writecount < size) {
			perror(out);
			fclose(f);
			fclose(g);
			return -1;
		}
	}
	fclose(f);
	fclose(g);
	delete [] buff;
	return 0;
}

int make_rom (int argc, char **argv) {
	char tmp[TMPFNAMESIZE];
	char *inputfile;
	char *outputfile;
	if (argc != 2) {
		fprintf(stderr, "vmipstool: --make-rom takes 2 arguments\n");
		usage();
		return 1;
	}
	sprintf (tmp, "/tmp/vmipstool-%d-1", (int) getpid ());
	inputfile = argv[0];
	outputfile = argv[1];
	if (!program_available (objcopyname))
		return config_error ("objcopy", objcopyname);
	if (echo_and_run_l (objcopyname, "-O", "binary", inputfile, tmp, NULL) != 0) {
		fprintf(stderr, "vmipstool: Error creating binary image of program.  Aborting.\n");
		return 1;
	}
	if (verbose)
		printf ("dd if=%s of=%s bs=%ld conv=sync > /dev/null 2>&1\n",
				tmp, outputfile, pagesz);
	if (copy_with_padded_blocks(tmp, outputfile, pagesz) != 0) {
		fprintf(stderr,
				"vmipstool: Error copying program image to ROM file.  Aborting.\n");
		return 1;
	}
	if (verbose)
		printf("rm %s\n", tmp);
	unlink(tmp);
	return 0;
}

/********************************* Byte-swapping ******************************/

int swap_words (int argc, char **argv) {
	char *inputfname = argv[0];
	char *outputfname = argv[1];
	FILE *inputfp = fopen(inputfname, "rb");
	if (!inputfp) {
		perror(inputfname);
		return -1;
	}
	FILE *outputfp = fopen(outputfname, "wb");
	if (!outputfp) {
		perror(outputfname);
		return -1;
	}
	char buff[4], buff2[4];
	int readcount, writecount;
	while ((readcount = fread(buff, 1, 4, inputfp)) > 0) {
		if (readcount < 4) {
			fprintf(stderr,
					"%s: warning: file does not end on a word boundary\n",
					inputfname);
			for (int i = readcount; i < 3; i++)
				buff[i] = 0;
		}
		buff2[0] = buff[3];
		buff2[1] = buff[2];
		buff2[2] = buff[1];
		buff2[3] = buff[0];
		if ((writecount = fwrite(buff2, 1, 4, outputfp)) != 4) {
			perror(outputfname);
			break;
		}
	}
	fclose(inputfp);
	fclose(outputfp);
	return 0;
}

/******************************** Everything else *****************************/

int compile (int argc, char **argv) {
	if (!program_available (ccname)) return config_error ("gcc", ccname);
	return echo_and_run_lv (ccname, endianflag, "-mno-abicalls",
				"-fno-pic", NULL, argv, NULL);
}

int assemble (int argc, char **argv) {
	if (!program_available (ccname)) return config_error ("gcc", ccname);
	return echo_and_run_lv (ccname, "-c", "-x", "assembler-with-cpp",
				endianflag, "-fno-pic", NULL, argv, NULL);
}

int preprocess (int argc, char **argv) {
	if (!program_available (ccname)) return config_error ("gcc", ccname);
	return echo_and_run_lv(ccname, "-E", NULL, argv, NULL);
}

int disassemble (int argc, char **argv) {
	char *inputfile = argv[0];
	if (!program_available (objdumpname))
		return config_error ("objdump", objdumpname);
	return echo_and_run_l (objdumpname, "--disassemble", endianflag,
                           inputfile, NULL);
}

int disassemble_rom (int argc, char **argv) {
	char *inputfile = argv[0];
	if (!program_available (objdumpname))
		return config_error ("objdump", objdumpname);
	return echo_and_run_l (objdumpname, "--disassemble-all", "--target=binary",
                           endianflag, "-m", "mips", inputfile, NULL);
}

int disassemble_word (int argc, char **argv) {
  EndianSelfTester est;
  Disassembler d (est.host_is_big_endian(), stdout);
  if (argc != 2) {
    fprintf (stderr, "Error: must supply PC and INSTR\n");
    return 1;
  }
  uint32 pc = strtoul(argv[0], NULL, 0);
  uint32 instr = strtoul(argv[1], NULL, 0);
  d.disassemble (pc, instr);
  return 0;
}

int version (int argc, char **argv) {
	printf("vmipstool (VMIPS) %s\n", VERSION);
	return 0;
}

int help (int argc, char **argv) {
	usage();
	return 0;
}

/******************************** Options handling ****************************/

int (*handler) (int argc, char **argv);
char **handler_argv = 0;
int handler_argc = -1;

class VmipstoolOptions : public Options {
public:
  VmipstoolOptions () : Options () {
  }
  virtual ~VmipstoolOptions () { }

  virtual void usage (char *argv0)
  {
    printf ("usage:\n"
            " %s [ VT-FLAGS ] MODE MODE-ARGS\n"
            "Each MODE takes its own MODE-ARGS:\n"
            "  MODE:              MODE-ARGS:\n"
            "  --compile          [ FLAGS ] FILE.c -o FILE.o\n"
            "  --preprocess       [ FLAGS ] FILE\n"
            "  --assemble         [ FLAGS ] FILE.s -o FILE.o\n"
            "  --link             [ FLAGS ] FILE1.o ... FILEn.o -o PROG\n"
            "  --make-rom         PROG PROG.rom\n"
            "  --disassemble-rom  PROG.rom\n"
            "  --disassemble-word PC INSTR\n"
            "  --disassemble      PROG (or FILE.o)\n"
            "  --swap-words       INPUT OUTPUT\n"
            "\n"
            "VT-FLAGS may include:\n"
            " --help         Display this help message and exit.\n"
            " --version      Display the version of vmipstool and exit.\n"
            " --verbose      Echo commands as they are run.\n"
            " --dry-run      Don't actually run anything; use with --verbose.\n"
            " --ld-script=T  Use T as ld script (instead of default script);\n"
            "                use with --link.\n"
            "\n" "Report bugs to <vmips@dgate.org>.\n", argv0);
  }

  virtual void process_options (int argc, char **argv)
  {
    /* Get options from defaults. */
    process_defaults ();

    /* Get default name of user's config file */
    char user_config_filename[PATH_MAX] = "~/.vmipsrc";

    /* Process command line */
    bool read_system_config_file = true;
    std::vector<std::string> command_line_options;

    int i;
    for (i = 1; (!handler) && (i < argc); ++i)
      {
        if (strcmp (argv[i], "--version") == 0)
          {
            print_package_version ("vmipstool", VERSION);
            exit (0);
          }
        else if (strcmp (argv[i], "--help") == 0)
          {
            usage (argv[0]);
            exit (0);
          }
        else if (strcmp (argv[i], "-o") == 0)
          {
            if (argc <= i + 1)
              error_exit ("The -o flag requires an argument. Try %s --help",
                           argv[0]);
            command_line_options.push_back (argv[i + 1]);
            ++i;
          }
        else if (strcmp (argv[i], "-F") == 0)
          {
            if (argc <= i + 1)
              error_exit ("The -F flag requires an argument. Try %s --help",
                           argv[0]);
            strcpy (user_config_filename, argv[i + 1]);
            ++i;
          }
        else if (strcmp (argv[i], "-n") == 0)
          {
            read_system_config_file = false;
          }
        else if (strcmp (argv[i], "--verbose") == 0)
          {
            verbose = 1;
          }
        else if (strcmp (argv[i], "--dry-run") == 0)
          {
            dryrun = 1;
          }
        else if (strncmp (argv[i], "--ld-script=", 12) == 0)
          {
            strcpy (ldscript_full_path, &argv[i][12]);
            ldscript_error = !can_read_file (ldscript_full_path);
          }
        else if (strcmp (argv[i], "--compile") == 0)
          {
            handler = compile;
          }
        else if (strcmp (argv[i], "--assemble") == 0)
          {
            handler = assemble;
          }
        else if (strcmp (argv[i], "--preprocess") == 0)
          {
            handler = preprocess;
          }
        else if (strcmp (argv[i], "--link") == 0)
          {
            handler = do_link;
          }
        else if (strcmp (argv[i], "--make-rom") == 0)
          {
            handler = make_rom;
          }
        else if (strcmp (argv[i], "--disassemble-rom") == 0)
          {
            handler = disassemble_rom;
          }
        else if (strcmp (argv[i], "--disassemble-word") == 0)
          {
            handler = disassemble_word;
          }
        else if (strcmp (argv[i], "--disassemble") == 0)
          {
            handler = disassemble;
          }
        else if (strcmp (argv[i], "--swap-words") == 0)
          {
            handler = swap_words;
          }
        else                    /* Unrecognized arg. */
          {
            std::cerr << "Unrecognized arg: " << argv[i] << "\n";
            usage (argv[0]);
            exit (1);
          }
      }
    if (!handler)
      {
        std::cerr << "Error: Missing mode arg.\n";
        usage (argv[0]);        /* Arg list was incomplete somehow. */
        exit (1);
      }
    handler_argc = argc - i;
    handler_argv = argv + i;

    /* Get options from system configuration file */
    if (read_system_config_file)
      process_options_from_file (SYSTEM_CONFIG_FILE);

    /* Get options from user configuration file */
    tilde_expand (user_config_filename);
    process_options_from_file (user_config_filename);

    /* Process -o options saved from command line, above. */
    for (std::vector<std::string>::iterator i =
         command_line_options.begin (), e = command_line_options.end ();
         i != e; ++i)
      process_one_option (i->c_str ());
  }
};

int
main (int argc, char **argv)
{
  /* Find the default ld script. */
  ldscript_error = !can_read_default_ldscript ();

  /* Get the command line options. */
  VmipstoolOptions *opt = new VmipstoolOptions;
  opt->process_options (argc, argv);
  setup_paths (opt->option ("bigendian")->flag,
               opt->option ("mipstoolprefix")->str);

  /* Run the appropriate handler. */
  if (!handler)
    fatal_error ("handler was not set");
  if (handler_argc == -1)
    fatal_error ("handler_argc was not set");
  if (!handler_argv)
    fatal_error ("handler_argv was not set");
  return handler (handler_argc, handler_argv);
}
